File: zeroMQ/python/mt-req-rep/req.py
#!/usr/bin/python 
# -*- coding: utf-8 -*-
#
# Copyright (c) 2013 ASMlover. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list ofconditions and the following disclaimer.
#
#    notice, this list of conditions and the following disclaimer in
#  * Redistributions in binary form must reproduce the above copyright
#    the documentation and/or other materialsprovided with the
#    distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import zmq
import sys
import random


if __name__ == '__main__':
  if (len(sys.argv) < 2):
    print 'arguments error ...'
    exit()

  ctx = zmq.Context()
  req = ctx.socket(zmq.REQ)
  req.connect('tcp://localhost:5555')
  
  random.seed()
  for i in range(int(sys.argv[1])):
    msg = 'request id %d' % random.randint(1, 1000)
    req.send(msg)
    print msg

    msg = req.recv()
    print 'recevied from reply => %s' % msg

  req.close()

File: zeroMQ/c/mt-req-rep/req.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <zmq.h> 


int 
main(int argc, char* argv[])
{
  void* ctx = zmq_ctx_new();
  void* req = zmq_socket(ctx, ZMQ_REQ);
  zmq_connect(req, "tcp://localhost:5555");

  do {
    int i, num_req;
    char buf[128];

    if (argc < 2) {
      fprintf(stderr, "arguments error ...\n");
      break;
    }
    num_req = atoi(argv[1]);

    srand((unsigned int)time(0));
    for (i = 0; i < num_req; ++i) {
      sprintf(buf, "request id {%d}", rand() % 1000);
      zmq_send(req, buf, strlen(buf), 0);
      fprintf(stdout, "send request : %s\n", buf);

      memset(buf, 0, sizeof(buf));
      zmq_recv(req, buf, sizeof(buf), 0);
      fprintf(stdout, "recevied from reply : %s\n", buf);
    }
  } while (0);

  zmq_close(req);
  zmq_ctx_destroy(ctx);

  return 0;
}
